src/BuildingRegistry.Api.Legacy/Infrastructure/Grb/Wfs/GrbWfsClient.cs

```diff
@@ -0,0 +1,134 @@
+namespace BuildingRegistry.Api.Legacy.Infrastructure.Grb.Wfs
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.IO;
+    using System.Linq;
+    using System.Net;
+    using System.Net.Http;
+    using System.Xml;
+    using System.Xml.Linq;
+    using System.Xml.XPath;
+    using NetTopologySuite.Geometries;
+    using NetTopologySuite.IO.GML2;
+
+    public interface IGrbWfsClient
+    {
+        IEnumerable<Tuple<Geometry, IReadOnlyDictionary<string, string>>> GetFeaturesInBoundingBox(GrbFeatureType featureType, Envelope boundingBox);
+    }
+
+    internal class GrbWfsClient : IGrbWfsClient
+    {
+        private readonly string _wfsUrl;
+        private readonly XmlNamespaceManager _namespaceManager;
+        private readonly GMLReader _gmlReader;
+
+        public GrbWfsClient(GrbWfsConfiguration configuration)
+        {
+            if (configuration == null)
+                throw new ArgumentNullException(nameof(configuration));
+
+            _wfsUrl = configuration.Url;
+
+            _namespaceManager = new XmlNamespaceManager(new NameTable());
+            _namespaceManager.AddNamespace("wfs", GrbWfsNameSpaces.Wfs);
+            _namespaceManager.AddNamespace("gml", GrbWfsNameSpaces.Gml);
+            _namespaceManager.AddNamespace("grb", GrbWfsNameSpaces.Grb);
+
+            _gmlReader = new GMLReader();
+        }
+
+        public IEnumerable<Tuple<Geometry, IReadOnlyDictionary<string, string>>> GetFeaturesInBoundingBox(GrbFeatureType type, Envelope boundingBox)
+        {
+            var featureName = GrbFeatureName.For(type);
+            var wfsRequest = CreateWfsRequest(featureName, boundingBox);
+
+            try
+            {
+                var response = wfsRequest
+                        .GetResponse()
+                        .GetResponseStream();
+
+                var wfsDoc = XDocument.Load(response);
+                if (wfsDoc.Root?.Name?.LocalName == null)
+                    throw new GrbWfsException("Invalid response");
+
+                if (wfsDoc.Root.Name.LocalName.Equals("ServiceExceptionReport", StringComparison.OrdinalIgnoreCase))
+                    throw new GrbWfsException(wfsDoc.ToString());
+
+                return wfsDoc
+                    .XPathSelectElements("/wfs:FeatureCollection/gml:featureMember", _namespaceManager)
+                    .Select(element => MapFeature(featureName, element));
+            }
+            catch (Exception exception)
+                when (exception is GrbWfsException)
```

Comment:
why not simply `throw` instead of making a new exception containing the other one?

Reply:
```csharp
catch (Exception exception) when (exception is GrbWfsException)
{
    throw;
}
```

Reply:
because somehow the `not` got deleted in the process

should be
```catch (Exception exception) when (!(exception is GrbWfsException))```

any exception here gets wrapped in a GrbWfsException, if it's already the correct type there is no point catching & rethrowing it